### ISSUE TITLE
Changing route doesn't scroll to top in the new page

### ISSUE THREAD

**Issue by netlander** [angular/angular#7791]
When navigating from one route to another then scrolling down a bit and navigate back to the previous route, the page remains at the same scroll position.

**Current behavior**

Navigate from one view to another then scroll down a bit and navigate back to the previous view. the page remains at the same scroll position.

**Expected/desired behavior**

When navigating back to another route the page should be scrolled to the top.

**Comment by zoechi** [comment]
dup of #6595 ?

**Comment by mgol** [comment]
> dup of #6595 ?

I don't think so.

**Comment by zoechi** [comment]
#6946 shows a workaround

**Comment by xak2000** [comment]
>**Expected/desired behavior**
>
>When navigating back to another route the page should be scrolled to the top.

It isn't expected behaviour for me. Expected behaviour for me is:

When navigating back to another route the page should be scrolled to position at which it was before going forward. This is the behaviour currently implemented in all browsers when navigating between normal pages (not SPA). But I never seen this behaviour in any implementation of angular1 router.

It is very annoying when you scroll down some list, click on list item to see details page, scrolling down details, then click browser `Back` button and not see the list page in exactly the same state as it was before.
- If it not scrolls at all: you just end up in random scroll position on the list page.
- If it scrolls to the top: it's better, but you end up always at top of the list page, while in any "normal" site (not SPA) you end up at scroll position at which you were been before going to detail page.

The github, for example, implements this correctly: Open issues list and scroll down, then open any issue details, then click `Back` browser button. You will end up at exactly the same scroll position at which you were been before opening issue details page.

It will be very cool if angular2 router able to do it. It is really a killer feature I never seen in any other routers.

**Comment by sulandr** [comment]
may be would be better ta have some attribute or param to change this default scroll behaviour?

**Comment by andyrue** [comment]
Is there a recommended method for doing this in RC.4? The suggested workaround in #6946 no longer works.

**Comment by cortopy** [comment]
@andyrue I don't think this would be recommended but I posted an update on #6946

**Comment by mhamel06** [comment]
In my case the `window.scrollTo(0, 0)` trick wouldn't work because I actually need to scroll to the top of an overflowing router-outlet. If you have a similar issue this is what I did as a workaround:

```
@Component({
  template: '<router-outlet (activate)="onActivate($event, outlet)" #outlet></router-outlet>,
})
export class MyParentComponent {
  onActivate(e, outlet){
    outlet.scrollTop = 0;
  }
}
```

**Comment by Ariix** [comment]
Any solution? None of the proposed solutions work.

**Comment by cortopy** [comment]
@Ariix wouldn't that scroll after every time change detection runs?

**Comment by Ariix** [comment]
@Cortopy Yes, yes it does. So still looking

**Comment by robinkedia** [comment]
Any update or milestone for this?

**Comment by naveedahmed1** [comment]
Any update on this? Its annoying when you scroll down some list, click a list item to see details (which is on another route) you have to scroll up. If the next details page's length is small and the list page is very long and by chance you clicked the last item in the list. It would give an impression that there aren't any contents on details page (as the user has to scroll up to view contents).

**Comment by zoechi** [comment]
@naveedahmed1 https://github.com/angular/angular/issues/6595#issuecomment-244232725

**Comment by naveedahmed1** [comment]
@zoechi I believe that comment is about URL with hash, in my case I don't' have any hash:

List page has a url:
 /list

details page has a url:
 /details/123

**Comment by zoechi** [comment]
@naveedahmed1 you can still call `scrollIntoView()` or `scrollTop=0`

**Comment by robinkedia** [comment]
@zoechi - Can you share the full code? i have parent child scenario where routes for child is changing.

**Comment by sime** [comment]
My work around involves the ElementRef class.

```js
  constructor(
    private route: ActivatedRoute,
    private service: GuideService,
    private router: Router,
    private element: ElementRef
  ) {
    this.scrollUp = this.router.events.subscribe((path) => {
      element.nativeElement.scrollIntoView();
    });
  }

```

**Comment by naveedahmed1** [comment]
How about this?

```
import { Component, OnInit} from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';

@Component({
    moduleId: module.id,
    selector: 'app',
    templateUrl: './app.component.html'
})

export class AppComponent implements OnInit {
    constructor(private router: Router) {}

    ngOnInit() {
        this.router.events.subscribe((evt) => {
            if (!(evt instanceof NavigationEnd)) {
                return;
            }

            var scrollToTop = window.setInterval(function () {
                var pos = window.pageYOffset;
                if (pos > 0) {
                    window.scrollTo(0, pos - 20); // how far to scroll on each step
                } else {
                    window.clearInterval(scrollToTop);
                }
            }, 16); // how fast to scroll (this equals roughly 60 fps)
        });
    }
}
```

**Comment by SmaugTheGreat** [comment]
router.events.filter(event => event instanceof NavigationEnd).subscribe(event => {
            window.scroll(0, 0);
        });

This seems to remember the scroll position (in Chrome), but I have no idea why.

**Comment by daco** [comment]
@SmaugTheGreat Yes, that does work! But how is that even possible that it remembers the last scroll position?

**Comment by naveedahmed1** [comment]
Its strange,  when using @SmaugTheGreat solution i.e. window.scroll(0, 0); it remembers scroll position.
But when using it window.scroll in window.setInterval, it doesn't remember scroll position.

**Comment by fxck** [comment]
why doesn't NavigationEnd contain params, query params etc of the route? only string?

**Comment by shanehickeylk** [comment]
A mixture of Smaugs solution and Guilherme Meireles on stackoverflow seems to work for me, as well as assigning the subscription to a variable so that I could unsubscribe in ngOnDestroy to avoid memory leakage:

`import {Component} from '@angular/core';
import {Router, NavigationEnd} from '@angular/router';
import {Subscription} from 'rxjs/Rx';

@Component({
    selector: ‘my-app’,
    template: `<router-outlet></router-outlet>`
})

export class AppComponent {
    routerSubscription: Subscription;

    constructor(private _router: Router) { }

    ngOnInit() {
        this.routerSubscription = this._router.events
            .filter(event => event instanceof NavigationEnd)
            .subscribe(event => {
                document.body.scrollTop = 0;
            });
    }

    ngOnDestroy() {
        this.routerSubscription.unsubscribe();
    }
}
`

**Comment by shanehickeylk** [comment]
A mixture of Smaugs solution above and Guilherme Meireles on stackoverflow does the trick for me. Also unsubscribing in ngOnDestroy to avoid leakage. 

    import { Component } from '@angular/core'; 
    import { Router, NavigationEnd } from '@angular/router'; 
    import { Subscription } from 'rxjs/Rx';

    @Component({
        moduleId: module.id, 
        selector: ‘my-app’,
        template: `<router-outlet></router-outlet>`
    })

    export class AppComponent {
        routerSubscription: Subscription;

        constructor(private router: Router) { }

        ngOnInit() {
            this.routerSubscription = this.router.events
                .filter(event => event instanceof NavigationEnd)
                .subscribe(event => {
                    document.body.scrollTop = 0;
                });
        }

        ngOnDestroy() {
            this.routerSubscription.unsubscribe();
        }
    }

**Comment by gowthamrodda** [comment]
use 'ng2-page-scroll' this module. As a quick response, i am posting one of my component. This one worked for me.

import {Component, OnInit,  Inject } from '@angular/core';
import {Router, Route, NavigationEnd}            from '@angular/router';
import { AppSharedService} from '../app.service';
import { DOCUMENT } from '@angular/platform-browser';
import { Subscription } from 'rxjs/Rx';
import { PageScrollService, PageScrollInstance }  from 'ng2-page-scroll';

@Component({
  selector: 'app-resources',
  templateUrl: './resources.component.html',
  styleUrls: ['./resources.component.scss']
})
export class ResourcesComponent {

  routerSubscription: Subscription;

  constructor(private router: Router, private SharedService: AppSharedService, private pageScrollService: PageScrollService,   @Inject(DOCUMENT) private document: any,) {
    let pageScrollInstance: PageScrollInstance = PageScrollInstance.simpleInstance(this.document, '#app-give');
    this.pageScrollService.start(pageScrollInstance);

  }


}

**Comment by spock123** [comment]
@shanehickeylk  
I believe you only need to unsubscribe because you specifically saves the subscription into a variable.

If you just called router.events..... .subscribe() directly, there is no subscription to unsubscribe.

**Comment by shanehickeylk** [comment]
@spock123 
I believe that the subscription is still "alive" even when it is not assigned to a variable. This is how angular knows an event has been triggered. The reason for assigning it to a variable is so that it can be accessed later and unsubscribed.

Here is a good article by Brian Love, where he goes into some additional detail on the topic:
http://brianflove.com/2016/12/11/anguar-2-unsubscribe-observables/

**Comment by spock123** [comment]
Aight, thanks.. need to rewrite some production  asap ,😎😎

**Comment by JoniJnm** [comment]
Maybe you want go to top only in root router changes (not in children, because you can load routes with lazy load in f.e. a tabset)

Another way to solve it:

```html
<router-outlet (deactivate)="onDeactivate()"></router-outlet>
```

```ts
onDeactivate() {
    document.body.scrollTop = 0;
}
```

**Comment by HauntedSmores** [comment]
Just want to report that I see this issue whenever I use a route resolver. Looks like if has to do with the timing of a component being destroyed/loaded in the router. When I use a resolver, seems the component loads really fast, too fast, and its like the router outlet is never truly empty or something so the scroll position gets saved.

**Comment by josh81** [comment]
I solved this issue by adding autoscroll="true" in the ui-view

**Comment by iwish0** [comment]
josh81, what you said is  for angular 1, not angular 2

**Comment by kormic** [comment]
@JoniJnm maybe using the renderer would be better?

`<router-outlet (deactivate)="onDeactivate()"></router-outlet>`

`import { Component, Renderer } from '@angular/core';`
  
`export class AppComponent {

  constructor(private renderer: Renderer) {
  }

  onDeactivate() {
    this.renderer.setElementProperty(document.body, "scrollTop", 0);
  }
}`

**Comment by mtpultz** [comment]
I added this to my `app-component::ngOnInit` to get around scrolling issues after routing where native scrolling on the window or document body wasn't working due to the content being in the sidenav content container.  In case this helps anyone temporarily until beta.4.

```
this.router
  .events
  .filter(event => event instanceof NavigationEnd)
  .subscribe(() => {
      const contentContainer = document.querySelector('.mat-sidenav-content') || this.window;
      contentContainer.scrollTo(0, 0);
  });
```

**Comment by xe4me** [comment]
I created a little service that handles this : 

[Tutorial](https://medium.com/@miladhi/always-scroll-to-top-of-the-page-with-this-simple-angular-plugin-8e6eec487108)

[Source](https://github.com/xe4me/scrollstore)

**Comment by Splaktar** [comment]
@kormic using the `Renderer` is a good idea! However, `Renderer` is deprecated, so `Renderer2` should be used now. Also `setElementProperty()` was changed to `setProperty()` in `Renderer2`.

I combined this with @kormic and @mtpultz's workarounds and this seems to be working fairly well for me atm:

```
    this.router.events.filter(event => event instanceof NavigationEnd).subscribe(() => {
      this.renderer.setProperty(document.body, 'scrollTop', 0);
    });
```

**Comment by ysus** [comment]
@Splaktar  i have a angular universal app. how to avoid the server exception.
the 3 way works but  i have a server message, :

window.scrollTo(0,0); //window not defined.
this.element.nativeElement.scrollIntoView(false); //scrollIntoView not defined
this.renderer.setProperty(document.body,'scrollTop',0); //document not defined

**Comment by ysus** [comment]
After an hours of work finally found a solution  for server side rendering
```
if (isPlatformBrowser(this.platformId)) {
                var scrollingElement = this.document.scrollingElement || this.document.documentElement;
                this.renderer.setProperty(scrollingElement ,'scrollTop',0);
            }
```

**Comment by DavidFrahm** [comment]
FYI, anyone who cares about this probably also cares about preserving scroll position for previous views when navigating "Back". That is in a separate issue #10929.

**Comment by JohnGalt1717** [comment]
My question is how is this combined with #hash anchors not supported out of the box and not known after like 10 iterations of routers in Angular?

1. When I click on a link it should scroll to the top.
2. If I click on a link and it has a # it should go to that page and scroll to it, or if it's the same page scroll down to it.
3. If I hit back or forward, it should go to where I was.

This isn't hard. I thought this was solved in like 1992.  Someone at Google please get into a stand-up and knock heads. This needs to be fixed yesterday. I shouldn't have to write hacks just to do what every browser since Lynx has been able to do.

**Comment by alanhg** [comment]
`app.component.ts`，add following code
this.router.events.filter(event => event instanceof NavigationEnd).subscribe((e: NavigationEnd) => {
            document.body.scrollTop = 0;
        });

**Comment by JohnGalt1717** [comment]
That amongst other things doesn't work properly with # notation on pages etc.  This needs to be fixed properly by the Angular team in the root code per the requirements above.

**Comment by padrian** [comment]
check this out it helped me;

[https://stackoverflow.com/questions/39601026/angular-2-scroll-to-top-on-route-change](url)

**Comment by MohamadAtieh** [comment]
This really needs to be the expected behaviour in SPAs. We are trying a workaround for our production app and it really supposed to be out of the box solution.

**Comment by jayoungers** [comment]
I would agree with most of the prior comments: since my application isn't in Production yet I'm holding off hoping this will be implemented without any workarounds needed.

**Comment by Azmos** [comment]
The autoscroll no longer works in Firefox :-/
55.0.1 (64 bit)

Chrome works as expected though.

**Comment by ysus** [comment]
this works for me
`let number = window.scrollY || window.pageYOffset || document.body.scrollTop + (document.documentElement && document.documentElement.scrollTop || 0)`

**Comment by dockleryxk** [comment]
My team has been using [what Google uses](https://github.com/angular/angular/blob/master/aio/src/app/shared/scroll.service.ts) on angular.io. Just make a service and inject it like usual. Then, on ngAfterViewInit on each page you want this behavior, just call this.[scroll service variable name].scrollToTop(). Finally, you'll need to add this to the top of <body> in index.html: `<div id="top-of-page"></div>`

**Comment by Pinguinmann** [comment]
Thank you in first place for all your workaround, but are you serious? 
I have to agree with xak2000 this is **not** expected behavior. 
I mean we have DI and Data-Binding, Pipes with a beast framework and stuff, but we have to workaround to scroll to top when the route changes? 
Sorry but this can not be a solution for me...

**Comment by ByronBeleris** [comment]
I just used 
`this.router.navigate(['myNextPage']).then(() => {
      window.scrollTo(0, 0)
    });`

and it worked.

**Comment by JohnGalt1717** [comment]
Again, that's isn't what it's supposed to do.

If you go to a regular old website and hit the back button it goes back AND keeps the exact position on the page where you clicked from so you're brought back to EXACTLY where you left off.

Angular needs to have this built in and it needs to just work. No hacking, no providing a bad estimation of the functionality. The real deal working exactly as a user would expect automatically.

**Comment by daviskoko** [comment]
Simply add this in your main component
```
constructor(private router: Router) {
    this.router.events.subscribe(
      () => window.scrollTo(0, 0)
    );
  }
```

**Comment by Hulva** [comment]
This works for me:
```
<router-outlet (deactivate)="onDeactivate()"></router-outlet>
onDeactivate() {
    window.scrollTo(0, 0);
}
```

**Comment by crysislinux** [comment]
@JohnGalt1717  what you see for old websites is handled by the browser. it's quite different for a framework like angular.  angular has to make use of javascript to handle the problem, which may be impossible to get a consistent and clean solution. All the other frameworks have the same problem. The only easy way to solve the problem out there is hide the element instead of destroy them when switching routes, then the browser will do the restore scroll position thing.

**Comment by JohnGalt1717** [comment]
@crysislinux  It's actually pretty trivial. In the history buffer where they intercept the back button and renavigate they also need to store the scroll position and when back /forward is hit, they scroll to that position after the ngOnInit() is complete.

**Comment by fxck** [comment]
except that you can't exactly do it on ngOnInit, since not all (say async) data might be rendered yet

**Comment by JohnGalt1717** [comment]
since ngOnInit allows async (promises) you can. If the developer doesn't do it right that's not your problem that's theirs for doing it wrong.

**Comment by jsgoupil** [comment]
I hope that you guys thought of restoring every single scrollable div I have on the page that has disappeared, that might load data asynchronously, that includes images, after I rotated my device in landscape.

There is nothing trivial about this... I am surprised nobody came up with a plugin so far. You would become famous instantly. In the mean time, just scroll to the top :-)

**Comment by bobrosoft** [comment]
Best solution I came with from usability side of it (at least in Chrome) is to do window.scroll manually right in places where I do **forward** navigation in that way:
```
this.router.navigate(['/payment']).then(() => {
  window.scroll(0, 0); // should be here, in promise
});
```
In that case Chrome correctly remembers scroll position if I navigate back and resets position if I navigate forward (but yes, you need to do it manually everywhere, or wrap into some kind of own method for navigation, not found another solution).

**Comment by tramel-woodard** [comment]
Just broadcasting that @kormic 's solution importing the `Renderer` object from the `@angular/core` library triggered by the `onDeactivate()` function within the **app.component.ts** file is still working for my Angular 4 project, built with angular-cli@4.0.0.

**Comment by kampageddon** [comment]
This work for me:
[https://stackoverflow.com/questions/39601026/angular-2-scroll-to-top-on-route-change](https://stackoverflow.com/questions/39601026/angular-2-scroll-to-top-on-route-change)

In your component:
```
import { Component, OnInit } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';

@Component({
    selector: 'my-app',
    template: '<ng-content></ng-content>',
})
export class MyAppComponent implements OnInit {
    constructor(private router: Router) { }

    ngOnInit() {
        this.router.events.subscribe((evt) => {
            if (!(evt instanceof NavigationEnd)) {
                return;
            }
            window.scrollTo(0, 0)
        });
    }
}
```

**Comment by tommyc38** [comment]
I upgraded to Angular 5 and none of these worked.  In fact, it seemed like none of my DOM calls worked at all (e.g. window.scrollTo(0, 0) ).

**Comment by ericmartinezr** [comment]
Will be solved by https://github.com/angular/angular/pull/20030

**Comment by KarthikSamyak** [comment]
Simple solution would be adding window.scrollTo(0,0) inside ngOnInit on the required component.
```javascript
ngOnInit() {
     window.scrollTo(0, 0);
}

**Comment by tommyc38** [comment]
@ericmartinezr thank you for the update.  I'm using Angular Material (MatSidenav specifically) which is why calls to window and body weren't working. (e.g. they weren't scrollable). Everything worked once I called it on the appropriate DOM element (e.g. `.mat-sidenav-content`).  I just assumed that elements `body` and `window` would be scrollable since they were parents of `.mat-sidenav-content`.

**Comment by raysuelzer** [comment]
> When navigating back to another route the page should be scrolled to position at which it was before  going forward. This is the behaviour currently implemented in all browsers when navigating between > normal pages (not SPA). But I never seen this behaviour in any implementation of angular1 router.

@xak2000 raises a great point, but the behavior I am seeing is that the scroll position is mainted regardless if it is a "back" action, or a navigation to an entirely new page. I'm fine with keeping the scroll position for back actions, but seems like a bug for navigating to a new page.

**Comment by hesampour** [comment]
thank @naveedahmed1 ; it worked very nice

**Comment by stephengardner** [comment]
Lord forgive me.  It's 3:00am, I have to be at work tomorrow morning and I'm sitting here trying to figure out how probably the most basic concept on a framework's 9th year in existence.

**Comment by peterpeterparker** [comment]
same as @bobrosoft scrolling when the promise resolve was the best solution for me

furthermore to chrome, it works in safari too

**Comment by KhaderMurtaja** [comment]
@naveedahmed1  thanks ^_^

**Comment by marticrespi** [comment]
window.scrollTo() works, but what's happen if you use universal where any DOM object exists?

**Comment by mgol** [comment]
@marticrespi There's no concept of scroll position on the server so just wrap the `scrollTo` code in a `isPlatformBrowser` check.

**Comment by sal-vader** [comment]
Here's my solution...


		  export class AppComponent implements OnInit {
		  isPopState = false;
		  
		  constructor(private router: Router, private locStrat: LocationStrategy) { }
		  
		  ngOnInit(): void {
			this.locStrat.onPopState(() => {
			  console.log('onPopState is running');
			  this.isPopState = true;
			});
			
			this.router.events.subscribe(event => {
			  // Scroll to top if accessing a page, not via browser history stack
			  if (event instanceof NavigationEnd && !this.isPopState) {
				window.scrollTo(0, 0);
				this.isPopState = false;
			  }
			  
			  // Ensures that isPopState is reset
			  if (event instanceof NavigationEnd) {
				this.isPopState = false;
			  }
			});
		  }
		}

**Comment by sridhark158** [comment]
it's worked nicely thank you ;)

**Comment by markgoho-EDT** [comment]
This is still broken in rc4

**Comment by maplion** [comment]
+1

**Comment by jongood01** [comment]
Waiting on a fix for this - you shouldn't have to have window.scrollTo anywhere as the 'normal' and expected behaviour of a webpage should be the default. All of the solutions suggested here scroll you to the top of the page when you click the back button which is not desired. Anyone have an idea on how to do this until it's fixed?

**Comment by hesampour** [comment]
the above solution works fine in many devices but as I checked it does not work on mobile devices. the following solution is what I Used to solve this issue in different kind of devices

`    if (isPlatformBrowser(this.platformId)) {
    if (window.screen.width <= 1024) {
     this.mobile = true;
    }

          this.routerSubscription = this.router.events
            .filter(event => event instanceof NavigationEnd)
            .subscribe(event => {
              if (this.mobile) {
                const element = document.querySelector('#scrollId1');
                element.scrollIntoView();
              } else {
                window.scrollTo(0, 0);
              }

            });
  }

`

**Comment by lucatardi** [comment]
In my case the above solutions weren't working, I checked on the console and I found out that the scrollTo() method of the window wasn't working, so I found another solution, instead of bringing the user at the position 0,0 of the page I used the scrollIntoView() method, bringing the top component of the page (in my case app-website-nav) into a view, and, It works!!!! now every time that the URL changes the view will go at the top of the page where my app-website-nav component is!
here the code, simple and works:
`
import { Router, NavigationEnd } from '@angular/router';
private router: Router,

constructor(private router: Router)

ngOnInit()
   {
     this.router.events.subscribe((evt) => {
                 if (!(evt instanceof NavigationEnd)) {
                     return;
                 }
                 document.getElementsByTagName("app-website-nav")[0].scrollIntoView();
             });
   }
`

**Comment by mkhan004** [comment]
for smooth scrolling with router change you can use 
```
  constructor(private router: Router) {}

  ngOnInit() {
    this.router.events.subscribe(() =>
      window.scrollTo({
        top: 0,
        left: 0,
        behavior: 'smooth'
      })
    );
  }
```

**Comment by peterpeterparker** [comment]
@mkhan004 interesting but may I ask, `behavior` is not browser native right (https://www.w3schools.com/jsref/met_win_scrollto.asp)? 

therefore are you using a special lib/polyfill which override `window.scrollTo` or something else?

**Comment by mgol** [comment]
@peterpeterparker It is native browser behavior. Word of advice - don't use W3Schools, it is known for imprecise or just plain wrong information. It's better to use MDN; a relevant page here: https://developer.mozilla.org/en-US/docs/Web/API/Window/scrollTo

**Comment by peterpeterparker** [comment]
@mgol thx 👍

**Comment by MartinMa** [comment]
Worth mentioning: `behavior: 'smooth'` is not supported by IE or Edge.

Current versions of Chrome and Firefox seem to work though.

**Comment by mgol** [comment]
@MartinMa That's a good point; the relevant Can I Use page is https://caniuse.com/#feat=css-scroll-behavior

Note that even just omitting `behavior: 'smooth'` will still break IE/Edge/Safari as they don't support the object syntax to `scrollTo`, just the positional one: `scrollTo(x, y)`. To support both of them with Chrome/Firefox getting smooth scroll and others non-smooth scroll, you can use the following logic (with the desired `x`/`y` values):
```js
if ('scrollBehavior' in document.documentElement.style) {
  window.scrollTo({
    behavior: 'smooth',
    left: x,
    top: y,
  });
} else {
  window.scrollTo(x, y);
}
```

**Comment by trotyl** [comment]
Closed by https://github.com/angular/angular/pull/20030

**Comment by jasonaden** [comment]
Closing based on being fixed as above.

**Comment by Johnyoat** [comment]
use `window.scrollTo(0,0);` in ngOnit. So it should look like this 
`ngOnInit() {
    window.scrollTo(0,0);
  }`

**Comment by javid-abd** [comment]
just have a look at Angular 6.1. It was fixed, and probably in a future release, it will be default

**Comment by burredis** [comment]
I **used scrollIntoView(false)** 

`<router-outlet (activate)='onActivate($event)'></router-outlet>`

`onActivate(e) {
    document.getElementsByClassName('user-notification')[0].scrollIntoView(false);
  }`

`('user-notification')[0]` is the element that I used to scroll top. `window.scrollTo(0,0)`, `document.body.scrollTo(0,0)` didn't work. I found this way to workaround this issue.

I hope help :)

**Comment by rajkumargadhavi** [comment]
here is the best solution!!!

import { Component } from '@angular/core';
import { Router, NavigationEnd } from '@angular/router';

@Component({
  selector: 'sv-app',
  template: '<router-outlet></router-outlet>'
})
export class AppComponent {

  constructor(private router: Router) {

    router.events
      .filter(event => event instanceof NavigationEnd)
      .subscribe((event: NavigationEnd) => {
        window.scroll(0, 0);
      });

  }
}

**Comment by JamieFarrelly** [comment]
A workaround is to put this in application.run() - at least in an old version of Angular. Might help someone!

```
var interval = setInterval(function() {
     if (document.readyState === 'complete') {
       $window.scrollTo(0, 0);
       clearInterval(interval);
     }
}, 10);

**Comment by Toub** [comment]
@JamieFarrelly an official solution has been [released in angular 6.1](https://github.com/angular/angular/blob/48d7205873f13c9a8ea417364d59e23b46478c79/CHANGELOG.md)
> router: implement scrolling restoration service (#20030) (49c5234), closes #13636 #10929 #7791 #6595

@see documentation:
- https://angular.io/api/router/ExtraOptions#scrollPositionRestoration
- https://angular.io/api/common/ViewportScroller

**Comment by PixelShoot** [comment]
@odahcam I'm using MatDrawerContainer and it works form with ViewChild                                              `@ViewChild('scroll') myScrollContainer:MatDrawerContainer; 
this.myScrollContainer.scrollable.getElementRef().nativeElement.scrollTop= 0;`

**Comment by odahcam** [comment]
Yeah, I'm able to scroll the container, but how to make the router automatically scroll it for me? (like it would with viewport)

**Comment by PixelShoot** [comment]
@odahcam 

```
  <router-outlet (activate)="onActivate($event)"></router-outlet>
```
```
onActivate(event){
    this.myScrollContainer.scrollable.getElementRef().nativeElement.scrollTop= 0;
}
```

**Comment by odahcam** [comment]
Actually I'm using something like that, but I think the router should have some scrollable container configuration, beucase it tries to scroll by the viewport when you enable the `scrollPositionRestoration: 'top'` option and it won't work. 

Your code does the job as mine does too, but both of us are unable to take profit from the native router functionality that does the scrolling job for us. To do that, we would need to tell the router to use our custom scrollable container instead of the viewport.

I noticed that the Router makes use of a guy called [`ViewportScroller`](https://angular.io/api/common/ViewportScroller) and, as the [docs](https://angular.io/api/router/ExtraOptions#scrollPositionRestoration) says, you should inject that guy in your module or component if you wan't to customize the scroll behavior:

```typescript
class AppModule {
 constructor(router: Router, viewportScroller: ViewportScroller, store: Store<AppState>) {
   router.events.pipe(filter(e => e instanceof Scroll), switchMap(e => {
     return store.pipe(first(), timeout(200), map(() => e));
   }).subscribe(e => {
     if (e.position) {
       viewportScroller.scrollToPosition(e.position);
     } else if (e.anchor) {
       viewportScroller.scrollToAnchor(e.anchor);
     } else {
       viewportScroller.scrollToPosition([0, 0]);
     }
   });
 }
}
```

That makes me think that the only way to integrate something like a fullscreen `cdkContainer` to the router scroll restoration would be implement my own `ViewportScroller` and alter the injection token that Angular uses for the router to inject my scroller instead of the default.

Unfortunately the docs for this matter is a little bit inexistent, so things are even more difficult.

**Comment by vishwackh** [comment]
Angular 6.1 and later:

RouterModule.forRoot(quoteToolRoutes, {scrollPositionRestoration: 'enabled'})

**Comment by vishwackh** [comment]
Angular 6.1 and later:

RouterModule.forRoot(routes, {scrollPositionRestoration: 'enabled'})

**Comment by Splaktar** [comment]
@odahcam I've linked your comments to a related Angular Material issue: https://github.com/angular/material2/issues/4280. It would be great if you could provide a Stackblitz example and some additional information there.

**Comment by odahcam** [comment]
@Splaktar all right, posted.

**Comment by Baathus** [comment]
> Angular 6.1 and later:
> 
> RouterModule.forRoot(routes, {scrollPositionRestoration: 'enabled'})

To others looking for a simple and clean solution for keeping the scroll position on back navigation and scroll to top on forward navigations, this is the solution. Will be default Angular behaviour in the future (described by the angular specifications: https://angular.io/api/router/ExtraOptions)

**Comment by maximelafarie** [comment]
> > Angular 6.1 and later:
> > RouterModule.forRoot(routes, {scrollPositionRestoration: 'enabled'})
> 
> To others looking for a simple and clean solution for keeping the scroll position on back navigation and scroll to top on forward navigations, this is the solution. Will be default Angular behaviour in the future (described by the angular specifications: https://angular.io/api/router/ExtraOptions)

I would like to add a half thumb up! 😅 

If you have the following `ExtraOptions`:

```typescript
// ...
const routerOptions: ExtraOptions = {
  useHash: false,
  anchorScrolling: 'enabled',
  scrollPositionRestoration: 'enabled'
};
// ...
```
in order to enable anchor links, it'll always go on top, even if the anchor is at the bottom of the page.

But if you aren't using anchor navigation, it's an awesome fix! 😉
